domain/application/service/application.go

```diff
@@ -1289,7 +1303,15 @@ func (s *Service) SetApplicationStatus(
 		return internalerrors.Errorf("encoding workload status: %w", err)
 	}
 
-	return s.st.SetApplicationStatus(ctx, applicationID, encodedStatus)
+	if err := s.st.SetApplicationStatus(ctx, applicationID, encodedStatus); err != nil {
+		return internalerrors.Capture(err)
+	}
+
+	if status == nil {
+		return nil
+	}
+
+	return s.statusHistory.RecordStatus(ctx, applicationNamespace.WithID(applicationID.String()), *status)
```

Comment:
We should annotate these so that if/when it fails, we can clearly see from the error message that it's "just" status history. 
EDIT: we do annotate for unit status history
Actually, I'm of the mind that errors writing status history should not be fatal. Really, failing here does not impact the juju model and its operation. Certainly in 3.6 status history was implemented in the Done() method of a model operation and was not fatal if it failed. We could have a helper method MaybeRecordStatusHistory which takes the same args as the above plus the recorder itself, and if the operation fails, logs a warning.

Reply:
Agreed. I'll fix them accordingly.